lib/node_modules/@stdlib/string/base/slice-code-points/lib/main.js

```diff
@@ -0,0 +1,124 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2025 The Stdlib Authors.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+// MODULES //
+
+var max = require( '@stdlib/math/base/special/max' );
+
+
+// VARIABLES //
+
+var RE_UTF16_LOW_SURROGATE = /[\uDC00-\uDFFF]/; // TODO: replace with stdlib pkg
+var RE_UTF16_HIGH_SURROGATE = /[\uD800-\uDBFF]/; // TODO: replace with stdlib pkg
+
+
+// MAIN //
+
+/**
+* Slices a string based on Unicode code points.
+*
+* @param {string} str - input string
+* @param {integer} start - the `ith` Unicode code point to start a slice (inclusive)
+* @param {integer} end - the `jth` Unicode code point to end a slice (exclusive)
+* @throws {TypeError} must provide exactly 3 arguments
+* @returns {string} output string
+*
+* @example
+* var out = sliceCodePoints( 'Hello 👋 World', 1, 3 );
+* // returns 'el'
+*
+* out = sliceCodePoints( '👋👋👋', 1, 2 );
+* // returns '👋'
+*/
+function sliceCodePoints( str, start, end ) {
+	var totalCodePoints;
+	var codePoints;
+	var len;
+	var ch1;
+	var ch2;
+	var idx;
+	var ch;
+	var i;
+
+	if ( arguments.length !== 3 ) {
+		throw new TypeError( 'invalid invocation. Must provide a valid string, starting code point index, and ending code point index.' );
+	}
+
+	len = str.length;
+	if ( len === 0 ) {
+		return '';
+	}
+
+	totalCodePoints = 0;
+	for ( i = 0; i < len; i++ ) {
+		ch1 = str[ i ];
+		totalCodePoints += 1;
+
+		if ( RE_UTF16_HIGH_SURROGATE.test( ch1 ) && i < len - 1 ) {
+			ch2 = str[ i + 1 ];
+			if ( RE_UTF16_LOW_SURROGATE.test( ch2 ) ) {
+				i += 1;
+			}
+		}
+	}
+
+	if ( start < 0 ) {
+		start = max( totalCodePoints + start, 0 );
+	}
+	if ( end < 0 ) {
+		end = max( totalCodePoints + end, 0 );
```

Comment:
FTR, this looks correct. Just unfortunate that we need to perform two passes given that we need to know that total number of code points before we can determine the end code point index.